otter/test/convergence/test_gathering.py

```diff
@@ -218,9 +235,101 @@ class GetScalingGroupServersTests(SynchronousTestCase):
     """
 
     def setUp(self):
-        """Save basic reused data."""
-        self.req = (ServiceType.CLOUD_SERVERS, 'GET',
-                    'servers/detail?limit=100')
+        self.now = datetime(2010, 5, 31)
+        self.servers1 = [{'id': 'a', 'a': 'b'}, {'id': 'b', 'b': 'c'}]
+        self.servers2 = [{'id': 'd', 'd': 'e'}]
+        self.freeze = compose(set, map(freeze))
+
+    def _invoke(self):
+        return get_scaling_group_servers(
+            'tid', 'gid', self.now, cache_class=Cache,
+            all_as_servers=intent_func("all-as"),
+            all_servers=intent_func("alls"))
+
+    def _test_no_cache(self, empty):
+        current = [] if empty else self.servers1
+        sequence = SequenceDispatcher([
+            ("cachegstidgid", lambda i: (object(), None)),
+            (("all-as",), lambda i: {} if empty else {"gid": current}),
+            (("cacheistidgid", self.now, current, True), noop)])
+        with sequence.consume():
+            disp = ComposedDispatcher([sequence, base_dispatcher])
+            self.assertEqual(
+                sync_perform(disp, self._invoke()), current)
+
+    def test_no_cache(self):
+        """
+        If cache is empty then current list of servers are returned and added
+        to the cache
+        """
+        self._test_no_cache(False)
+        self._test_no_cache(True)
+
+    def _test_old_case(self, last_update, as_srvs=True, cur_srvs=True):
+        exp_last_update = datetime(2010, 5, 1)
+        as_servers = self.servers1
+        current = self.servers2
+        servers = (as_srvs and as_servers or []) + (cur_srvs and current or [])
+        sequence = SequenceDispatcher([
+            ("cachegstidgid", lambda i: (object(), last_update)),
+            (("all-as", exp_last_update),
+             lambda i: {'gid': as_servers} if as_srvs else {}),
+            (("all-as",), lambda i: {"gid": current} if cur_srvs else {}),
+            (("cacheistidgid", self.now, servers, True), noop)])
+        with sequence.consume():
+            disp = ComposedDispatcher([sequence, base_dispatcher])
+            self.assertEqual(
+                sync_perform(disp, self._invoke()), servers)
+
+    def test_old_cache(self):
+        """
+        If cache is older than 30 days then servers returned are got by getting
+        current list and changes since last 30 days. The cache is updated with
+        this list
+        """
+        dt = datetime(2010, 3, 1)
+        self._test_old_case(dt)
+        self._test_old_case(dt, False, True)
+        self._test_old_case(dt, True, False)
+
+    def test_from_cache(self):
+        """
+        If cache is < 30 days old then servers returned are merge of
+        changes since the cache time
+        """
+        asmetakey = "rax:autoscale:group:id"
+        cache = [
+            {'id': 'a', 'metadata': {asmetakey: "gid"}},
+            {'id': 'b', 'metadata': {asmetakey: "gid"}}]
```

Comment:
To test that cached servers get removed if in the changes, the metadata has been changed, could we also add a server with id `d` to the cache?